fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/NewQueryBuilder.java

```diff
@@ -181,8 +225,48 @@ public Select buildQuery(Class<?> resourceType, FHIRSearchContext searchContext)
                 new Object[] { resourceType.getSimpleName(), searchContext.getSearchParameters() });
 
         final SearchQuery domainModel;
-
-        if (searchContext.hasSortParameters()) {
+        if (Resource.class.equals(resourceType)) {
+            // Whole-system search
+            if (allSearchParmsAreGlobal(searchContext.getSearchParameters())) {
+                // Can do a filter query against global tables
+                SearchWholeSystemFilterQuery wholeSystemFilterQuery = new SearchWholeSystemFilterQuery();
+                for (SortParameter sp: searchContext.getSortParameters()) {
+                    wholeSystemFilterQuery.add(new DomainSortParameter(sp));
+                }
+                if (searchContext.getSearchResourceTypes() != null) {
+                    // The _type parameter was specified, so we need to filter the
+                    // query by resource type. Add an extension to the domain model
+                    // with the specified resource type IDs.
+                    addResourceTypeExtension(wholeSystemFilterQuery, searchContext.getSearchResourceTypes());
+                }
+                domainModel = wholeSystemFilterQuery;
+            } else {
+                // Not all search parameters are global (values indexed into global values tables).
+                // Need to do old-style UNION'd query against all resource types.
+                List<String> resourceTypes = searchContext.getSearchResourceTypes();
+                if (resourceTypes == null) {
+                    // The _type parameter was not specified, so we need to generate a list
+                    // of all supported resource types for our UNION query.
+                    resourceTypes = new ArrayList<>(this.identityCache.getResourceTypeNames());
+                    resourceTypes.remove("Resource");
```

Comment:
Remove will be a linear scan each time. Although perhaps making this a set would be slower. I guess we can create this list at bootstrap time (in the servlet listener) when we add proper support for limited resource types in the schema.